src/serializers/type_serializers/union.rs

```diff
@@ -113,94 +116,42 @@ fn to_python(
-            Ok(v) => return Ok(v),
-            Err(err) => errors.push(err),
-        }
-    }
-
-    // If extra.check is SerCheck::Strict, we're in a nested union
-    if extra.check != SerCheck::Strict && retry_with_lax_check {
-        new_extra.check = SerCheck::Lax;
-        for comb_serializer in choices {
-            if let Ok(v) = comb_serializer.json_key(key, &new_extra) {
-                return Ok(v);
-            }
-        }
-    }
 
-    // If extra.check is SerCheck::None, we're in a top-level union. We should thus raise the warnings
-    if extra.check == SerCheck::None {
-        for err in &errors {
-            extra.warnings.custom_warning(err.to_string());
-        }
-    }
-    // Otherwise, if we've encountered errors, return them to the parent union, which should take
-    // care of the formatting for us
-    else if !errors.is_empty() {
-        let message = errors.iter().map(ToString::to_string).collect::<Vec<_>>().join("\n");
-        return Err(PydanticSerializationUnexpectedValue::new_err(Some(message)));
-    }
-    infer_json_key(key, extra)
-}
-
-#[allow(clippy::too_many_arguments)]
-fn serde_serialize<S: serde::ser::Serializer>(
-    value: &Bound<'_, PyAny>,
-    serializer: S,
-    include: Option<&Bound<'_, PyAny>>,
-    exclude: Option<&Bound<'_, PyAny>>,
-    extra: &Extra,
-    choices: &[CombinedSerializer],
-    retry_with_lax_check: bool,
-) -> Result<S::Ok, S::Error> {
-    let py = value.py();
-    let mut new_extra = extra.clone();
-    new_extra.check = SerCheck::Strict;
-    let mut errors: SmallVec<[PyErr; SMALL_UNION_THRESHOLD]> = SmallVec::new();
-
-    for comb_serializer in choices {
-        match comb_serializer.to_python(value, include, exclude, &new_extra) {
-            Ok(v) => return infer_serialize(v.bind(py), serializer, None, None, extra),
-            Err(err) => errors.push(err),
-        }
-    }
-
-    // If extra.check is SerCheck::Strict, we're in a nested union
-    if extra.check != SerCheck::Strict && retry_with_lax_check {
-        new_extra.check = SerCheck::Lax;
-        for comb_serializer in choices {
-            if let Ok(v) = comb_serializer.to_python(value, include, exclude, &new_extra) {
-                return infer_serialize(v.bind(py), serializer, None, None, extra);
+    if let Some(tag) = discriminator_value {
+        let tag_str = tag.to_string();
+        if let Some(&serializer_index) = lookup.get(&tag_str) {
+            let selected_serializer = &choices[serializer_index];
+
+            match selector(selected_serializer, &new_extra) {
+                Ok(v) => return Some(v),
+                Err(_) => {
+                    if retry_with_lax_check {
+                        new_extra.check = SerCheck::Lax;
+                        if let Ok(v) = selector(selected_serializer, &new_extra) {
+                            return Some(v);
+                        }
+                    }
+                }
             }
         }
```

Comment:
I wonder, should `tagged_union_serialize` call into `union_serialize` here? I think it seems like the established pattern that if `tagged_union_serialize` couldn't select a serializer, then `union_serialize` is always the immediate follow-up.